### ISSUE TITLE
Possible improvement to AngularJS orderBy.js

### ISSUE THREAD

**Issue by OpenSiteMobile** [angular/angular.js#14881]
As best I understand, the orderBy.js file was improved in Angular version 1.5.7 to deal with "tiebreakers" using the getComparisonObject function:

"NOTE: We are adding an extra `tieBreaker` value based on the element's index."

But I also noticed that the doComparison function which "compares" the values, uses the "compare" function defined under the orderByFilter function. Since the orderByFilter function can use a developer supplied comparison function (which would probably not be aware of the comparison tiebreaker code), would it not be better to have the doComparison function written as this:

```
    function doComparison(v1, v2) {
      for (var i = 0, ii = predicates.length; i < ii; i++) {
        var result = compare(v1.predicateValues[i], v2.predicateValues[i]);
        if (result) {
          return result * predicates[i].descending * descending;
        }
      }

      return defaultCompare(v1.tieBreaker, v2.tieBreaker) * descending;
    }
```

...so that the tiebreaker feature is always available regardless of whether or not the developer knew about it.

Current behavior:

Would ingnore the tieBreaker code from getComparisonObject function if the developer supplies his/her own comparison function.

Expected behavior:

Although the use case where this makes a difference may be small, adding the above change would always force the "tieBreaker" code to work where applicable. The downside might be upexpected results from developers not aware of the new "tieBreaker" comparison feature when upgrading to this version of AngularJS.

Angular version 1.5.7 is the first version I noticed the code for "tieBreaker" and should be browser independent, ref. file:

https://github.com/angular/angular.js/blob/master/src/ng/filter/orderBy.js

**Comment by gkalpak** [comment]
> As best I understand, the orderBy.js file was improved in Angular version 1.5.7 to deal with "tiebreakers"

Yes and no.
Yes it was improved (adding support for custom comparators, improving docs, cleaning up the code a bit).
No, the tie-breaker feature was not introduced in 1.5.7. It has been there for a while. We just documented it more explicitly, because now that a custom comparator can be specified, the user needs to be aware of it.

--
> Since the orderByFilter function can use a developer supplied comparison function (which would probably not be aware of the comparison tiebreaker code)

That's why we documented it, so that the developer is aware :smiley:

--
> would it not be better to have the doComparison function written as this: [...]

Not really, because the developer might want to supply custom tie-breaker logic.

--
> Current behavior: Would ingnore the tieBreaker code from getComparisonObject function if the developer supplies his/her own comparison function.

It will not ignore it. It will be passed to the developer-supplied comparator if necessary.

---
You have a point though: If the custom comparator fails to determine the order by any predicate and it fails to determine the order by the tie-breaker (i.e. `compare(v1.tieBreaker, v2.tieBreaker)` returns 0), there might be non-determinism (depending on the browser).

To avoid this (but at the same time allow the developer to handle this themselves if they want to), I suggest we call the default comparator _only_ if the custom comparator can't break the tie.

I.e. change `return compare(v1.tieBreaker, v2.tieBreaker) * descending` in `doComparison()` with:

```js
return (compare(v1.tieBreaker, v2.tieBreaker) || defaultCompare(v1.tieBreaker, v2.tieBreaker)) * descending
```

<sub>(formatted in a nicer way :stuck_out_tongue: )</sub>

@OpenSiteMobile , would you like to take a stab at it?

**Comment by OpenSiteMobile** [comment]
Sorry about not replying sooner. I don't think you need the "compare()" in:
```
return (compare(v1.tieBreaker, v2.tieBreaker) || defaultCompare(v1.tieBreaker, v2.tieBreaker)) * descending;
```
...just the "defaultCompare", since most often it is not going to be a tie. If the developer writes his "compare" function correctly to cover a specific "tie" he is concerned about, then his function will never reach this return (compare() || defaultCompare()) value. If he doesn't, then the "defautCompare" function is used, which is most likely what he wants anyway.

**Comment by gkalpak** [comment]
> I don't think you need the "compare()" [...] just the "defaultCompare", since most often it is not going to be a tie.

Because ties are expected to be rare indeed, it doesn't mean that you don't need to compare the tie-breaker with the custom comparator (for the rare case that this is needed).

> If the developer writes his "compare" function correctly to cover a specific "tie" he is concerned about, then his function will never reach this return (compare() || defaultCompare()) value.

That is not true. The custom comparator does not know anything about the predicates, so if the comparator can't decide based on the current values, it shouldn't consider it a definitive tie, since a subsequent predicate might break the tie.
The only place where you know that no predicate was able to break the tie is _after_ you have tried them all; i.e. when comparing using the tie-breaker value.

> If he doesn't, then the "defautCompare" function is used, which is most likely what he wants anyway.

True but "most likely" is not good enough. There might be cases where the user wants total control and doesn't want as to tell them how to break their ties, because "we thought that is most likely what you wanted" :stuck_out_tongue: 

--
Based on the above, I still believe it is correct to give the user full control by passing them the tie-breaker if needed. Additionally, this is already public API and changing it would be a breaking change (for no real value imo).

**tl;dr**
I insist we should use both `compare` and `defaultCompare` :smile:

**Comment by austinoneil** [comment]
One thing that I'm thinking is the following:

1. Follow the algorithm as is.
2. Break the sorted array into an array of arrays where each second-level array is completely tied.
3. On each of those second-level arrays, do a defaultSort.
4. "Flatten" the array and return the flattened array.

Does this seem like a fool-proof plan?  Worst case scenario, it adds O(n log n) to the sort time, which sorting is generally done in O(n log n) anyway, so this should work.

*I am assuming that the plan is to, essentially, add a "tiebreaker" to the end of the sort for when all the sorting is done.*

**Comment by gkalpak** [comment]
@austinoneil, I don't 100% follow your suggestion, but on what aspect of the current implementation does it improve?

**Comment by austinoneil** [comment]
@gkalpak If, at some point in the future, Chrome, Firefox, and IE drift away on their sorting algorithms to no longer be stable algorithms (instead going for performance), there will still be some consistency across browsers.  Is my italicized assumption correct?

**Comment by gkalpak** [comment]
We already have a tie breaker (exactly for the purpose of ensuring consistent sorting results across different invocations or browsers).

In a nutshell, for each object in the input array we create a "tie-breaker" object based on its index in the array. If all predicates fail to sort two objects, we use the tie-breaker to determine their relative order.

--
What we are debating here is what should happen in the case where the user specifies a custom comparator that can't sort/distinguish between some elements (for any of the predicates).

Right now, be pass our generated tie-breaker objects to that custom comparator (as documented). If it still can't sort them, then we pass the tie-breaker objects to our built-in comparator, which we know can deterministically sort the generated tie-breaker objects.

What @OpenSiteMobile suggests is that we should skip the step where we pass the tie-breaker objects to the custom comparator and pass them to our built-in comparator directly.
I suggest that we should give the user full control by keeping both steps.

**Comment by austinoneil** [comment]
That's what my idea was.  Since the defaultComparator sorts between primitives in addition to complex objects, there will be sets of objects that are indistinguishable to the tiebreaker.  My plan is to break the array into arrays where each element is "indistinguishable" to the tiebreaker (i.e. the tiebreaker would consider them to be ordered no matter what order they're in).  We sort those arrays using `defaultCompare` and combine those newly-sorted lists and return that list.

**Comment by austinoneil** [comment]
I made a pull request for this, but honestly, I'm kind of skeptical of whether this feature would be worth the potential problems.  For one, the implementation I did would add a sort, which wouldn't increase the asymptotic tight bound of the time, but could still potentially have an impact.  In addition, it seems like the `OrderBy` filter is designed to use a stable sort, so if the user had their list in a certain order, that order could matter to the user and adding a `DefaultSort` would mess things up, which would force the user to add a predicate to their filter, which would clutter their code.

**Comment by BobChao87** [comment]
I'd be willing to do this. Coming up with a good test case sounds pretty interesting. Currently it would be accurate to say that all we want is to fall back on the default if the user-provided comparison fails, correct?

**Comment by gkalpak** [comment]
@BobChao87, that sounds about right. We essentially need what is described in the last part of https://github.com/angular/angular.js/issues/14881#issuecomment-231164555.

**Comment by BobChao87** [comment]
Obviously the code changes were trivial, but I've been struggling trying to write a test for this and was hoping to get a pointer in the right direction. Way I see it, I have two options for this particular change:

1. Generate a list and user comparison function that will fail to have a given sorting outcome besides by falling back to the built-in.
   * Issue: Since fallback is original order of the list, no changes can be perceived by a list in a stable sort environment. Possible some obscure list/comparator exists, but I've not found it.
2. Use a spy on the function on the built-in function.
   * Issue: Cannot connect the spy since the function is internal.
      * Possible Remedy: Mock the `orderBy` filter to get the spy installed.
         * Issue: We cannot mock the interface we are testing, since that will not result in the desired code being tested.

So... any ideas? Ways to accomplish either of the above or possible other tests to achieve the desired results?

**Comment by gkalpak** [comment]
@BobChao87, it is a bit of a hack, but if you use `reverse` + a custom comparator which always returns 0, then you should be able to tell the difference. E.g.:

```js
orderBy([1, 2], '', true, () => 0)
// => [1, 2] when the default comparator is not used (in most browsers)
// => [2, 1] when the default comparator is used (always)
```

Just to be (more) sure, you could also test that `orderBy([1, 2], '', false, () => 0) => [1, 2]` (to make sure that it is not the browser that passes the items in reverse order).

##
So, the test might look somthing like:

```js
it('should ...', function() {
  var list = [1, 2];
  var comparator = valueFn(0);

  expect(orderByFilter(list, '', false, comparator)).toEqual([1, 2]);
  expect(orderByFilter(list, '', true, comparator)).toEqual([2, 1]);
});
```
